cpp/src/spatial/point_in_polygon.cu

```diff
@@ -0,0 +1,233 @@
+            poly_offsets.size(),
+            poly_offsets.begin<cudf::size_type>(),
+            poly_ring_offsets.begin<cudf::size_type>(),
+            poly_points_x.begin<T>(),
+            poly_points_y.begin<T>(),
+            results->mutable_view().begin<cudf::size_type>()
+        );
+
+        return results;
+    }
+};
+
+} // anonymous namespace
+
+namespace cuspatial {
+
+namespace detail {
+
+std::unique_ptr<cudf::column>
+point_in_polygon(cudf::column_view const& test_points_x,
+                 cudf::column_view const& test_points_y,
+                 cudf::column_view const& poly_offsets,
+                 cudf::column_view const& poly_ring_offsets,
+                 cudf::column_view const& poly_points_y,
+                 cudf::column_view const& poly_points_x,
+                 rmm::mr::device_memory_resource* mr,
+                 cudaStream_t stream)
+{
+    return cudf::experimental::type_dispatcher(test_points_x.type(), point_in_polygon_functor(),
+                                               test_points_x,
+                                               test_points_y,
+                                               poly_offsets,
+                                               poly_ring_offsets,
+                                               poly_points_y,
+                                               poly_points_x,
+                                               mr,
+                                               stream);
+}
+
+}
+
+std::unique_ptr<cudf::column>
+point_in_polygon(cudf::column_view const& test_points_x,
+                 cudf::column_view const& test_points_y,
+                 cudf::column_view const& poly_offsets,
+                 cudf::column_view const& poly_ring_offsets,
+                 cudf::column_view const& poly_points_y,
+                 cudf::column_view const& poly_points_x,
+                 rmm::mr::device_memory_resource* mr)
+{
+    CUSPATIAL_EXPECTS(test_points_x.size() == test_points_x.size() and
+                      poly_points_x.size() == poly_points_y.size(),
+                      "All points must have both x and y values");
+
+    CUSPATIAL_EXPECTS(test_points_y.type() == test_points_x.type() and
+                      poly_points_x.type() == test_points_x.type() and
+                      poly_points_y.type() == test_points_x.type(),
+                      "All points much have the same type for both x and y");
+
+    CUSPATIAL_EXPECTS(not test_points_x.has_nulls() &&
+                      not test_points_y.has_nulls(),
+                      "Test points must not contain nulls");
+
+    CUSPATIAL_EXPECTS(not poly_points_y.has_nulls() &&
+                      not poly_points_x.has_nulls(),
+                      "Polygon points must not contain nulls");
+
+    CUSPATIAL_EXPECTS(poly_offsets.size() <= (cudf::size_type) sizeof(cudf::size_type) * 8,
+                      "Number of polygons cannot exceed bitmap capacity (32 for cudf::size_type)");
+
```

Comment:
Again, I don't think size_type is appropriate for a bitmask. Notably, your size calculation should be replaced with `std::numeric_limits<mask_type>::digits`, and if you use a signed int type you will get 31 instead of 32.

Reply:
Afaik the signedness of the type does not matter because we're doing bit manipulation, and we're limited to signed types for cudf columns. I've changed this from `size_type` to `int32_t`. The consumer should not be inspecting the value as if it were a number, but as a per-point hit-mask. Please correct me if an int32 cannot represent a 32 bit mask - I have no experience doing this in c++.

I will add test cases for 31, 32, and 33 poly inputs, depending on which can be supported

Reply:
Am, I'm seeing how _at the very least_ this is highly inconvenient. :disappointed:

Reply:
Going with `int32_t` and supporting only 31 polygons for now.